api/utils/route.go

```diff
@@ -0,0 +1,155 @@
+
+// WithCaseInsensitiveRouting enables special case-insensitivity behavior when enabled. The value passed
+// here should correspond to the CaseInsensitiveRouting field of cluster networking config.
+func WithCaseInsensitiveRouting(cir bool) SSHRouteOption {
+	return func(opts *sshRouteOptions) {
+		opts.caseInsensitive = cir
+	}
+}
+
+// SSHRouteMatcher is a helper used to decide if an ssh dial request should match
+// a given server. This is broken out of proxy.Router as a standalone helper in order
+// to let other parts of teleport easily find matching servers when generating
+// error messages or building access requests.
+type SSHRouteMatcher struct {
+	sshRouteOptions
+	targetHost     string
+	targetPort     string
+	ips            []string
+	matchServerIDs bool
+}
+
+// NewSSHRouteMatcher builds a new matcher for ssh routing decisions.
+func NewSSHRouteMatcher(host, port string, opts ...SSHRouteOption) SSHRouteMatcher {
+	var options sshRouteOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	_, err := uuid.Parse(host)
+	dialByID := err == nil || IsEC2NodeID(host)
+
+	ips, _ := net.LookupHost(host)
+
+	return SSHRouteMatcher{
+		sshRouteOptions: options,
+		targetHost:      host,
+		targetPort:      port,
+		ips:             ips,
+		matchServerIDs:  dialByID,
+	}
+}
+
+// sshServer is an interface describing the subset of the types.Server interface
+// required to make a routing decision.
+type sshServer interface {
+	GetName() string
+	GetHostname() string
+	GetAddr() string
+	GetUseTunnel() bool
+}
+
+func (m *SSHRouteMatcher) RouteToServer(server sshServer) bool {
+	// if host is a UUID or EC2 ID match only
+	// by server name and treat matches as unambiguous
+	if m.matchServerIDs && server.GetName() == m.targetHost {
+		return true
+	}
+
+	// if the server has connected over a reverse tunnel
+	// then match only by hostname.
+	if server.GetUseTunnel() {
+		return m.routeToHostname(server.GetHostname())
+	}
+
+	ip, nodePort, err := net.SplitHostPort(server.GetAddr())
+	if err != nil {
+		return false
+	}
+
+	if (m.targetHost == ip || m.routeToHostname(server.GetHostname()) || slices.Contains(m.ips, ip)) &&
+		(m.targetPort == "" || m.targetPort == "0" || m.targetPort == nodePort) {
+		return true
+	}
+
+	return false
+}
+
+// routeToHostname helps us perform a special kind of case-insensitive comparison. SSH certs do not generally
+// treat principals/hostnames in a case-insensitive manner. This is often worked-around by forcing all principals and
+// hostnames to be lowercase. For backwards-compatibility reasons, teleport must support case-sensitive routing by default
+// and can't do this. Instead, teleport nodes whose hostnames contain uppercase letters will present certs that include both
+// the literal hostname and a lowered version of the hostname, meaning that it is sane to route a request for host 'foo' to
+// host 'Foo', but it is not sane to route a request for host 'Bar' to host 'bar'.
```

Comment:
I'm thinking that maybe theis explained by this comment [here](https://github.com/gravitational/teleport/pull/29320/files#diff-23a8678e981b764265fa92e70b0b16d7cdc98c9df59b136609b694df3031c11aR386-R391), which I wasn't aware of when I wrote that.

```
    # Enables case-insensitivity for ssh dial requests. Setting this to true will allow
    # 'tsh ssh myserver' to route to host 'MyServer'. Dialing with an uppercase hostname
    # will still require an exact match. Openssh always lowercases hostnames, so setting
    # this to true is necessary if you want to use openssh to access nodes with uppercase
    # characters in their hostnames.
    case_insensitive_routing: false
```

In that case I'm confused why we need this asymmetry, where you can dial lowercase to uppercase, but not uppercase to lowercase, but then with OpenSSH you can dial uppercase to lowercase because OpenSSH automatically lowercases things. Wouldn't it be simpler to just make this matching logic entirely case sensitive/insensitive based on that setting? (Is there a backwards compatibility problem with doing that that I'm missing?)

Reply:
The asymmetry here is because ssh certificate validation is not case-insensitive, meaning that `tsh ssh Foo` getting routed to host `foo` fails at cert validation because `Foo` is not a valid principal on the cert presented by the node.  This PR adds a lowercase copy of all hostnames to agent certs, so hosts name `Foo` now present a cert that contains `Foo` and `foo` as valid principals, but does not actually change the ssh cert validation logic (I looked into changing actual cert validation logic, but it seemed like it'd require more mucking about with `crypto/ssh` internals than I'd like, and anyhow its not clear to me that ssh certs were ever really intended to be case insensitive, so it seems safer to use them as intended and just support lowercase hostnames everywhere, which is consistent with how openssh handles things).

_However_, if I can get this PR in by `v14.0.0` then it would be theoretically safe/allowable to start lowercasing all hostnames in `tsh` just like openssh already does with hostnames after `v15.0.0`.  We'd need to figure out a scheme for propagating the networking config to `tsh` in a performant way... maybe caching it on login or something.  Regardless, doing this form of weaker case-insensitivity now opens us up for future improvements later while also avoiding routing to agents whose certs will fail validation 100% of the time (which IMO is confusing).